src/wayland/handlers/screencopy/render.rs

```diff
@@ -326,39 +326,66 @@ pub fn render_workspace_to_buffer(
     let transform = output.current_transform();
     let common = &mut state.common;
 
-    let result = match &mut state.backend {
-        BackendData::Kms(kms) => {
-            let render_node = kms
-                .target_node_for_output(&output)
-                .unwrap_or(kms.primary_node.expect("No Software Rendering"));
-            let target_node = get_dmabuf(&buffer)
-                .ok()
-                .and_then(|dma| dma.node())
-                .unwrap_or(render_node);
-
-            let buffer_format = match buffer_type(&buffer) {
-                Some(BufferType::Dma) => Some(get_dmabuf(&buffer).unwrap().format().code),
-                Some(BufferType::Shm) => {
-                    with_buffer_contents(&buffer, |_, _, data| shm_format_to_fourcc(data.format))
-                        .unwrap()
-                }
-                _ => None,
-            };
-            let mut multirenderer = match kms.api.renderer(
-                &render_node,
-                &target_node,
-                buffer_format.unwrap_or(Fourcc::Abgr8888),
+    let renderer = match state.backend.offscreen_renderer(|kms| {
+        let render_node = kms.target_node_for_output(&output).or(kms.primary_node)?;
```

Comment:
Looking at this, I wonder if this line is responsible for the flickering we sometimes see in the overview mode for multi-gpu systems. The target node is the native one for direct scanout, but not necessarily a good render_node for offscreen rendering.

(This is more a note, than something actionable.)

Reply:
It could make sense just to use a `single_renderer` here, with the target node of the buffer.

Or I guess it would make sense if it used the same logic as deciding what GPU to render on screen, but targeting the buffer node instead of the output node?

Though all of this is suboptimal without https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/268, so we can be sure we actually know what node the capture buffer, and the buffers we need to composite, are actually on.